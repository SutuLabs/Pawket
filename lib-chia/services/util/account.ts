import { AccountEntity, CustomCat, TokenInfo } from "../../models/account";
import receive, { TokenPuzzleDetail, TokenPuzzleObserver } from "../crypto/receive";

export const DEFAULT_ADDRESS_RETRIEVAL_COUNT = 1;

export async function getAccountAddressDetails(
  account: AccountEntity,
  cats: CustomCat[],
  tokenInfo: TokenInfo,
  prefix: string,
  symbol: string,
  maxId: number | undefined = undefined,
  catModName: "cat_v1" | "cat_v2" = "cat_v2"
): Promise<TokenPuzzleDetail[]> {
  if (typeof maxId !== "number" || maxId <= 0) maxId = account.addressRetrievalCount;
  if (typeof maxId !== "number" || maxId <= 0) DEFAULT_ADDRESS_RETRIEVAL_COUNT;
  if (!account.key.privateKey) throw new Error("Private key cannot empty for non-pk account to get account addresses.");

  if (account.addressGenerated == maxId) {
    return account.addressPuzzles;
  }

  if (account.type == "PublicKey") throw new Error("Public Key accound cannot provide puzzle detail");

  account.addressPuzzles = await receive.getAssetsRequestDetail(
    account.key.privateKey,
    0,
    maxId,
    cats,
    tokenInfo,
    prefix,
    symbol,
    catModName
  );
  account.addressGenerated = maxId;
  return account.addressPuzzles;
}

export async function getAccountPuzzleObservers(
  account: AccountEntity,
  cats: CustomCat[],
  tokenInfo: TokenInfo,
  prefix: string,
  symbol: string,
  maxId: number | undefined = undefined,
  catModName: "cat_v1" | "cat_v2" = "cat_v2"
): Promise<TokenPuzzleObserver[]> {
  if (typeof maxId !== "number" || maxId <= 0) maxId = account.addressRetrievalCount;
  if (typeof maxId !== "number" || maxId <= 0) DEFAULT_ADDRESS_RETRIEVAL_COUNT;

  if (account.addressGenerated == maxId) {
    return account.type == "PublicKey" ? account.observePuzzles ?? [] : account.addressPuzzles;
  }

  if (account.type == "PublicKey") {
    if (!account.key.publicKey) throw new Error("public key cannot empty for pk account to get account addresses.");
    account.observePuzzles = await receive.getAssetsRequestObserver(
      account.key.publicKey,
      0,
      maxId,
      cats,
      tokenInfo,
      prefix,
      symbol,
      catModName
    );
    account.addressGenerated = maxId;
    return account.observePuzzles;
  } else {
    if (!account.key.privateKey) throw new Error("Private key cannot empty for non-pk account to get account addresses.");
    account.addressPuzzles = await receive.getAssetsRequestDetail(
      account.key.privateKey,
      0,
      maxId,
      cats,
      tokenInfo,
      prefix,
      symbol,
      catModName
    );
    account.addressGenerated = maxId;
    return account.addressPuzzles;
  }
}
