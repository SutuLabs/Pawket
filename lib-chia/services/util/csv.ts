export function csvToArray(text: string): string[][] {
  // https://stackoverflow.com/a/8497474
  // Return array of string values, or NULL if CSV string not well formed.
  function CsvLinetoArray(text: string) {
    const re_valid =
      /^\s*(?:'[^'\\]*(?:\\[\S\s][^'\\]*)*'|"[^"\\]*(?:\\[\S\s][^"\\]*)*"|[^,'"\s\\]*(?:\s+[^,'"\s\\]+)*)\s*(?:,\s*(?:'[^'\\]*(?:\\[\S\s][^'\\]*)*'|"[^"\\]*(?:\\[\S\s][^"\\]*)*"|[^,'"\s\\]*(?:\s+[^,'"\s\\]+)*)\s*)*$/;
    const re_value =
      /(?!\s*$)\s*(?:'([^'\\]*(?:\\[\S\s][^'\\]*)*)'|"([^"\\]*(?:\\[\S\s][^"\\]*)*)"|([^,'"\s\\]*(?:\s+[^,'"\s\\]+)*))\s*(?:,|$)/g;
    // Return NULL if input string is not well formed CSV string.
    if (!re_valid.test(text)) return null;
    const a = []; // Initialize array to receive values.
    text.replace(
      re_value, // "Walk" the string using replace with callback.
      function (m0, m1, m2, m3) {
        // Remove backslash from \' in single quoted values.
        if (m1 !== undefined) a.push(m1.replace(/\\'/g, "'"));
        // Remove backslash from \" in double quoted values.
        else if (m2 !== undefined) a.push(m2.replace(/\\"/g, '"'));
        else if (m3 !== undefined) a.push(m3);
        return ""; // Return empty string.
      }
    );
    // Handle special case of empty last value.
    if (/,\s*$/.test(text)) a.push("");
    return a;
  }

  const lines = text.split("\n");
  const ops = [];
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (line.trim() == "") continue;

    const carr = CsvLinetoArray(line);
    if (carr == null) {
      // console.error("malformat of csv: ", line);
      throw new Error("malformat of csv: " + line);
    }

    ops.push(carr);
  }
  return ops;
}
