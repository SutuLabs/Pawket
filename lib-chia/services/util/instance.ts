import { ModuleInstance } from "@chiamine/bls-signatures";
import loadBls from "@chiamine/bls-signatures";
import * as clvm_tools from "clvm_tools";

class InstanceSingleton {
  private static _instance: InstanceSingleton = new InstanceSingleton();

  BLS?: ModuleInstance;

  constructor() {
    if (InstanceSingleton._instance) {
      throw new Error("Error: Instantiation failed: Use InstanceSingleton.getInstance() instead of new.");
    }
    InstanceSingleton._instance = this;
  }

  public static getInstance(): InstanceSingleton {
    return InstanceSingleton._instance;
  }

  public async init(): Promise<void> {
    this.BLS = await loadBls();
    await clvm_tools.initialize();
  }
}

export const Instance = InstanceSingleton.getInstance();
