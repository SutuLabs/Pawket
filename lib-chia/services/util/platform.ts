export function isIos(): boolean {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  if ((<any>window).cordova && (<any>window).cordova.platformId == "ios") return true;
  return false;
}
