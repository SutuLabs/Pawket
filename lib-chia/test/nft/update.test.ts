import { Instance } from "../../services/util/instance";
import { testUpdateNft } from "./functions";

beforeAll(async () => {
  await Instance.init();
});

test("Update Nft", async () => {
  await testUpdateNft(0n);
});
