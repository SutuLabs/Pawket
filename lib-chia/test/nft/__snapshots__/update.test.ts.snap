// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update Nft: analysis 1`] = `
Object {
  "coin": Object {
    "amount": 1n,
    "parent_coin_info": "0xdf5592ae5aead18a9a98d76c3d54b292511d9b230bdfd29975d7b4d6f35a05d5",
    "puzzle_hash": "0x72f4131e47360c7b5d04481de13c320a368ad89cbf6e4b0dfad0e6661bf6e978",
  },
  "didOwner": "",
  "hintPuzzle": "0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10",
  "launcherId": "0x9335e9ef78d3750c2368c02da0c50f4b6d7ac679e8346cdf8db324a100087c47",
  "launcherPuzzleHash": "0xeff07522495060c066f66f32acc2a77e3a3e737aca8baea4d1a64ea4cdc13da9",
  "metadata": Object {
    "imageHash": "e55788ca78f11f3e5524f35267b845f49d79840ba593e4411ba30b70f4d4d3e7",
    "imageUri": "https://storage.pawket.app/ipfs/QmRDDsfYLZvJ4aHRTYPnWyNvK8aJ2iFX8D8EaCJN7ncLZR?filename=good.xch.jpg",
    "licenseHash": "",
    "licenseUri": "",
    "metadataHash": "",
    "metadataUri": "",
    "serialNumber": "01",
    "serialTotal": "01",
  },
  "metadataUpdaterPuzzleHash": "0xfe8a4b4e27a2e29a4d3fc7ce9d527adbcaccbab6ada3903ccf3ba9a769d2d78b",
  "nextCoinName": "0x8b7aefab77c09eb90891558141c5c82737e395735aa59bf8b872267b2c4755a7",
  "nftOwnershipModHash": "0xc5abea79afaa001b5427dfa0c8cf42ca6f38f5841b78f9b3c252733eb2de2726",
  "nftStateModHash": "0xa04d9f57764f54a43e4030befb4d80026e870519aaa66334aef8304f5d0393c2",
  "p2InnerPuzzle": "(a (q 2 (q 2 (i 11 (q 2 (i (= 5 (point_add 11 (pubkey_for_exp (sha256 11 (a 6 (c 2 (c 23 ()))))))) (q 2 23 47) (q 8)) 1) (q 4 (c 4 (c 5 (c (a 6 (c 2 (c 23 ()))) ()))) (a 23 47))) 1) (c (q 50 2 (i (l 5) (q 11 (q . 2) (a 6 (c 2 (c 9 ()))) (a 6 (c 2 (c 13 ())))) (q 11 (q . 1) 5)) 1) 1)) (c (q . 0x97a9c3095a68a85f9cb8affeb9145e539e7388c37da435cd24e35493bc3ae4a2b7fdd0078a92eea6fa944c7722594050) 1))",
  "p2Owner": "7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10",
  "previousOwner": "()",
  "rawMetadata": "((117 \\"https://storage.pawket.app/ipfs/QmRDDsfYLZvJ4aHRTYPnWyNvK8aJ2iFX8D8EaCJN7ncLZR?filename=good.xch.jpg\\") (104 . 0xe55788ca78f11f3e5524f35267b845f49d79840ba593e4411ba30b70f4d4d3e7) (28021) (27765) (29550 . 1) (29556 . 1))",
  "royaltyAddress": "0x20c2755ec70535afc8a2b0408d47c30d47a1ce1232ccd3379030c3a51363e2b2",
  "singletonModHash": "0x7faa3253bfddd1e0decb0906b2dc6247bbc4cf608f58345d173adb63e8b47c9f",
  "tradePricePercentage": 200,
}
`;

exports[`Update Nft: spendbundle 1`] = `
Object {
  "aggregated_signature": "0xae2c529e2409b2a73558381a8992785d41a43e57a8152fd7e6b51d7635ddb5615fc8a393c1001174c397ace91c4e48b30286ddabf8a638c4ac76b89798940161f0c79284c6ce273c7eed289b941ef4b06bcec5ad9e7ef6da923d7eadc30cb5b3",
  "coin_spends": Array [
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0xdf5592ae5aead18a9a98d76c3d54b292511d9b230bdfd29975d7b4d6f35a05d5",
        "puzzle_hash": "0x72f4131e47360c7b5d04481de13c320a368ad89cbf6e4b0dfad0e6661bf6e978",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ffff18ff2fff3480ffff01ff04ffff04ff20ffff04ff2fff808080ffff04ffff02ff3effff04ff02ffff04ff05ffff04ffff02ff2affff04ff02ffff04ff27ffff04ffff02ffff03ff77ffff01ff02ff36ffff04ff02ffff04ff09ffff04ff57ffff04ffff02ff2effff04ff02ffff04ff05ff80808080ff808080808080ffff011d80ff0180ffff04ffff02ffff03ff77ffff0181b7ffff015780ff0180ff808080808080ffff04ff77ff808080808080ffff02ff3affff04ff02ffff04ff05ffff04ffff02ff0bff5f80ffff01ff8080808080808080ffff01ff088080ff0180ffff04ffff01ffffffff4947ff0233ffff0401ff0102ffffff20ff02ffff03ff05ffff01ff02ff32ffff04ff02ffff04ff0dffff04ffff0bff3cffff0bff34ff2480ffff0bff3cffff0bff3cffff0bff34ff2c80ff0980ffff0bff3cff0bffff0bff34ff8080808080ff8080808080ffff010b80ff0180ffff02ffff03ffff22ffff09ffff0dff0580ff2280ffff09ffff0dff0b80ff2280ffff15ff17ffff0181ff8080ffff01ff0bff05ff0bff1780ffff01ff088080ff0180ff02ffff03ff0bffff01ff02ffff03ffff02ff26ffff04ff02ffff04ff13ff80808080ffff01ff02ffff03ffff20ff1780ffff01ff02ffff03ffff09ff81b3ffff01818f80ffff01ff02ff3affff04ff02ffff04ff05ffff04ff1bffff04ff34ff808080808080ffff01ff04ffff04ff23ffff04ffff02ff36ffff04ff02ffff04ff09ffff04ff53ffff04ffff02ff2effff04ff02ffff04ff05ff80808080ff808080808080ff738080ffff02ff3affff04ff02ffff04ff05ffff04ff1bffff04ff34ff8080808080808080ff0180ffff01ff088080ff0180ffff01ff04ff13ffff02ff3affff04ff02ffff04ff05ffff04ff1bffff04ff17ff8080808080808080ff0180ffff01ff02ffff03ff17ff80ffff01ff088080ff018080ff0180ffffff02ffff03ffff09ff09ff3880ffff01ff02ffff03ffff18ff2dffff010180ffff01ff0101ff8080ff0180ff8080ff0180ff0bff3cffff0bff34ff2880ffff0bff3cffff0bff3cffff0bff34ff2c80ff0580ffff0bff3cffff02ff32ffff04ff02ffff04ff07ffff04ffff0bff34ff3480ff8080808080ffff0bff34ff8080808080ffff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff2effff04ff02ffff04ff09ff80808080ffff02ff2effff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff02ffff03ffff21ff17ffff09ff0bff158080ffff01ff04ff30ffff04ff0bff808080ffff01ff088080ff0180ff018080ffff04ffff01ffa07faa3253bfddd1e0decb0906b2dc6247bbc4cf608f58345d173adb63e8b47c9fffa09335e9ef78d3750c2368c02da0c50f4b6d7ac679e8346cdf8db324a100087c47a0eff07522495060c066f66f32acc2a77e3a3e737aca8baea4d1a64ea4cdc13da9ffff04ffff01ff02ffff01ff02ffff01ff02ff3effff04ff02ffff04ff05ffff04ffff02ff2fff5f80ffff04ff80ffff04ffff04ffff04ff0bffff04ff17ff808080ffff01ff808080ffff01ff8080808080808080ffff04ffff01ffffff0233ff04ff0101ffff02ff02ffff03ff05ffff01ff02ff1affff04ff02ffff04ff0dffff04ffff0bff12ffff0bff2cff1480ffff0bff12ffff0bff12ffff0bff2cff3c80ff0980ffff0bff12ff0bffff0bff2cff8080808080ff8080808080ffff010b80ff0180ffff0bff12ffff0bff2cff1080ffff0bff12ffff0bff12ffff0bff2cff3c80ff0580ffff0bff12ffff02ff1affff04ff02ffff04ff07ffff04ffff0bff2cff2c80ff8080808080ffff0bff2cff8080808080ffff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff2effff04ff02ffff04ff09ff80808080ffff02ff2effff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff02ffff03ff0bffff01ff02ffff03ffff09ff23ff1880ffff01ff02ffff03ffff18ff81b3ff2c80ffff01ff02ffff03ffff20ff1780ffff01ff02ff3effff04ff02ffff04ff05ffff04ff1bffff04ff33ffff04ff2fffff04ff5fff8080808080808080ffff01ff088080ff0180ffff01ff04ff13ffff02ff3effff04ff02ffff04ff05ffff04ff1bffff04ff17ffff04ff2fffff04ff5fff80808080808080808080ff0180ffff01ff02ffff03ffff09ff23ffff0181e880ffff01ff02ff3effff04ff02ffff04ff05ffff04ff1bffff04ff17ffff04ffff02ffff03ffff22ffff09ffff02ff2effff04ff02ffff04ff53ff80808080ff82014f80ffff20ff5f8080ffff01ff02ff53ffff04ff818fffff04ff82014fffff04ff81b3ff8080808080ffff01ff088080ff0180ffff04ff2cff8080808080808080ffff01ff04ff13ffff02ff3effff04ff02ffff04ff05ffff04ff1bffff04ff17ffff04ff2fffff04ff5fff80808080808080808080ff018080ff0180ffff01ff04ffff04ff18ffff04ffff02ff16ffff04ff02ffff04ff05ffff04ff27ffff04ffff0bff2cff82014f80ffff04ffff02ff2effff04ff02ffff04ff818fff80808080ffff04ffff0bff2cff0580ff8080808080808080ff378080ff81af8080ff0180ff018080ffff04ffff01a0a04d9f57764f54a43e4030befb4d80026e870519aaa66334aef8304f5d0393c2ffff04ffff01ffff75ffc06468747470733a2f2f73746f726167652e7061776b65742e6170702f697066732f516d5244447366594c5a764a346148525459506e57794e764b38614a326946583844384561434a4e376e634c5a523f66696c656e616d653d676f6f642e7863682e6a706780ffff68a0e55788ca78f11f3e5524f35267b845f49d79840ba593e4411ba30b70f4d4d3e7ffff826d7580ffff826c7580ffff82736e01ffff8273740180ffff04ffff01a0fe8a4b4e27a2e29a4d3fc7ce9d527adbcaccbab6ada3903ccf3ba9a769d2d78bffff04ffff01ff02ffff01ff02ffff01ff02ff26ffff04ff02ffff04ff05ffff04ff17ffff04ff0bffff04ffff02ff2fff5f80ff80808080808080ffff04ffff01ffffff82ad4cff0233ffff3e04ff81f601ffffff0102ffff02ffff03ff05ffff01ff02ff2affff04ff02ffff04ff0dffff04ffff0bff32ffff0bff3cff3480ffff0bff32ffff0bff32ffff0bff3cff2280ff0980ffff0bff32ff0bffff0bff3cff8080808080ff8080808080ffff010b80ff0180ff04ffff04ff38ffff04ffff02ff36ffff04ff02ffff04ff05ffff04ff27ffff04ffff02ff2effff04ff02ffff04ffff02ffff03ff81afffff0181afffff010b80ff0180ff80808080ffff04ffff0bff3cff4f80ffff04ffff0bff3cff0580ff8080808080808080ff378080ff82016f80ffffff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff2fffff01ff80ff808080808080808080ff0bff32ffff0bff3cff2880ffff0bff32ffff0bff32ffff0bff3cff2280ff0580ffff0bff32ffff02ff2affff04ff02ffff04ff07ffff04ffff0bff3cff3c80ff8080808080ffff0bff3cff8080808080ffff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff2effff04ff02ffff04ff09ff80808080ffff02ff2effff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff02ffff03ff5fffff01ff02ffff03ffff09ff82011fff3880ffff01ff02ffff03ffff09ffff18ff82059f80ff3c80ffff01ff02ffff03ffff20ff81bf80ffff01ff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff81dfffff04ff82019fffff04ff82017fff80808080808080808080ffff01ff088080ff0180ffff01ff04ff819fffff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff81dfffff04ff81bfffff04ff82017fff808080808080808080808080ff0180ffff01ff02ffff03ffff09ff82011fff2c80ffff01ff02ffff03ffff20ff82017f80ffff01ff04ffff04ff24ffff04ffff0eff10ffff02ff2effff04ff02ffff04ff82019fff8080808080ff808080ffff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff81dfffff04ff81bfffff04ffff02ff0bffff04ff17ffff04ff2fffff04ff82019fff8080808080ff8080808080808080808080ffff01ff088080ff0180ffff01ff02ffff03ffff09ff82011fff2480ffff01ff02ffff03ffff20ffff02ffff03ffff09ffff0122ffff0dff82029f8080ffff01ff02ffff03ffff09ffff0cff82029fff80ffff010280ff1080ffff01ff0101ff8080ff0180ff8080ff018080ffff01ff04ff819fffff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff81dfffff04ff81bfffff04ff82017fff8080808080808080808080ffff01ff088080ff0180ffff01ff04ff819fffff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff81dfffff04ff81bfffff04ff82017fff808080808080808080808080ff018080ff018080ff0180ffff01ff02ff3affff04ff02ffff04ff05ffff04ff0bffff04ff81bfffff04ffff02ffff03ff82017fffff0182017fffff01ff02ff0bffff04ff17ffff04ff2fffff01ff808080808080ff0180ff8080808080808080ff0180ff018080ffff04ffff01a0c5abea79afaa001b5427dfa0c8cf42ca6f38f5841b78f9b3c252733eb2de2726ffff04ffff0180ffff04ffff01ff02ffff01ff02ffff01ff02ffff03ff81bfffff01ff04ff82013fffff04ff80ffff04ffff02ffff03ffff22ff82013fffff20ffff09ff82013fff2f808080ffff01ff04ffff04ff10ffff04ffff0bffff02ff2effff04ff02ffff04ff09ffff04ff8205bfffff04ffff02ff3effff04ff02ffff04ffff04ff09ffff04ff82013fff1d8080ff80808080ff808080808080ff1580ff808080ffff02ff16ffff04ff02ffff04ff0bffff04ff17ffff04ff8202bfffff04ff15ff8080808080808080ffff01ff02ff16ffff04ff02ffff04ff0bffff04ff17ffff04ff8202bfffff04ff15ff8080808080808080ff0180ff80808080ffff01ff04ff2fffff01ff80ff80808080ff0180ffff04ffff01ffffff3f02ff04ff0101ffff822710ff02ff02ffff03ff05ffff01ff02ff3affff04ff02ffff04ff0dffff04ffff0bff2affff0bff2cff1480ffff0bff2affff0bff2affff0bff2cff3c80ff0980ffff0bff2aff0bffff0bff2cff8080808080ff8080808080ffff010b80ff0180ffff02ffff03ff17ffff01ff04ffff04ff10ffff04ffff0bff81a7ffff02ff3effff04ff02ffff04ffff04ff2fffff04ffff04ff05ffff04ffff05ffff14ffff12ff47ff0b80ff128080ffff04ffff04ff05ff8080ff80808080ff808080ff8080808080ff808080ffff02ff16ffff04ff02ffff04ff05ffff04ff0bffff04ff37ffff04ff2fff8080808080808080ff8080ff0180ffff0bff2affff0bff2cff1880ffff0bff2affff0bff2affff0bff2cff3c80ff0580ffff0bff2affff02ff3affff04ff02ffff04ff07ffff04ffff0bff2cff2c80ff8080808080ffff0bff2cff8080808080ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff3effff04ff02ffff04ff09ff80808080ffff02ff3effff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01ffa07faa3253bfddd1e0decb0906b2dc6247bbc4cf608f58345d173adb63e8b47c9fffa09335e9ef78d3750c2368c02da0c50f4b6d7ac679e8346cdf8db324a100087c47a0eff07522495060c066f66f32acc2a77e3a3e737aca8baea4d1a64ea4cdc13da9ffff04ffff01a020c2755ec70535afc8a2b0408d47c30d47a1ce1232ccd3379030c3a51363e2b2ffff04ffff018200c8ff0180808080ffff04ffff01ff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b092c7027d455e843b1226b757c1dd5e124e0e6ddab3549036a4684b410182fdfd1eee459eb76a4d6d73e94f90fbe6eb1dff018080ff018080808080ff018080808080ff01808080",
      "solution": "0xffffa09335e9ef78d3750c2368c02da0c50f4b6d7ac679e8346cdf8db324a100087c47ffa0f9294cfaf9977013d85114769de27e3ff026b396cc85232f530e6e422bd55beeff0180ff01ffffffff80ffff01ffff81e8ffff02ffff01ff04ffff04ffff02ffff03ffff22ff27ff3780ffff01ff02ffff03ffff21ffff09ff27ffff01826d7580ffff09ff27ffff01826c7580ffff09ff27ffff01758080ffff01ff02ff02ffff04ff02ffff04ff05ffff04ff27ffff04ff37ff808080808080ffff010580ff0180ffff010580ff0180ffff04ff0bff808080ffff01ff808080ffff04ffff01ff02ffff03ff05ffff01ff02ffff03ffff09ff11ff0b80ffff01ff04ffff04ff0bffff04ff17ff198080ff0d80ffff01ff04ff09ffff02ff02ffff04ff02ffff04ff0dffff04ff0bffff04ff17ff8080808080808080ff0180ff8080ff0180ff018080ffff759968747470733a2f2f6578616d706c652e636f6d2f612e6a706780ffff33ffa07ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10ff01ffffa07ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10808080ff8080808080",
    },
  ],
}
`;
